packages/meditrak-server/src/routes/mapOverlayGroups/assertMapOverlayGroupsPermissions.js

```diff
@@ -0,0 +1,213 @@
+        models,
+        mapOverlayRelation.child_id,
+      );
+      if (!result.result) {
+        return result;
+      }
+    }
+  }
+
+  if (mapOverlayGroupRelations.length) {
+    for (const mapOverlayRelation of mapOverlayRelations) {
+      const result = await hasMapOverlayGroupEditPermissions(
+        accessPolicy,
+        models,
+        mapOverlayRelation.child_id,
+      );
+      if (!result.result) {
+        return result;
+      }
+    }
+  }
+
+  return {
+    result: true,
+  };
+};
+
+export const assertMapOverlayGroupsGetPermissions = async (
+  accessPolicy,
+  models,
+  mapOverlayGroupId,
+) => {
+  const result = await hasMapOverlayGroupGetPermissions(accessPolicy, models, mapOverlayGroupId);
+  if (result.result) {
+    return true;
+  }
+
+  throw new Error(result.errorMessage);
+};
+
+export const assertMapOverlayGroupsEditPermissions = async (accessPolicy, models, mapOverlayId) => {
+  const result = await hasMapOverlayGroupEditPermissions(accessPolicy, models, mapOverlayId);
+
+  if (result.result) {
+    return true;
+  }
+
+  throw new Error(result.errorMessage);
+};
+
+const findPermittedMapOverlayGroupIds = (
+  relationByChildId,
+  permittedMapOverlayGroupRelations,
+  rootMapOverlayGroupId,
+) => {
+  const permittedMapOverlayGroupIds = [];
+
+  permittedMapOverlayGroupRelations.forEach(r => {
+    if (r.map_overlay_group_id === rootMapOverlayGroupId) {
+      return;
+    }
+
+    permittedMapOverlayGroupIds.push(r.map_overlay_group_id);
+  });
+
+  if (permittedMapOverlayGroupRelations.length === 0) {
+    return [];
+  }
+
+  const newRelations = permittedMapOverlayGroupIds
+    .map(id => relationByChildId[id])
+    .filter(r => !!r);
+
+  const nextPermittedMapOverlayGroupIds = findPermittedMapOverlayGroupIds(
+    relationByChildId,
+    newRelations,
+  );
+
+  return [...permittedMapOverlayGroupIds, ...nextPermittedMapOverlayGroupIds];
+};
+
+export const createMapOverlayGroupDBFilter = async (accessPolicy, models, criteria) => {
+  if (hasBESAdminAccess(accessPolicy)) {
+    return criteria;
+  }
+
+  const dbConditions = { ...criteria };
+
+  const allMapOverlayGroupRelations = await models.mapOverlayGroupRelation.find();
+  const relationByChildId = keyBy(allMapOverlayGroupRelations, 'child_id');
+
+  // Pull the list of map overlays we have access to,
+  // then pull the corresponding map overlay groups
+  const mapOverlaysFilter = createMapOverlayDBFilter(accessPolicy);
+  const permittedMapOverlays = await models.mapOverlay.find(mapOverlaysFilter);
+  const permittedMapOverlayGroupRelations = await models.mapOverlayGroupRelation.find({
+    child_id: permittedMapOverlays.map(mo => mo.id),
+    child_type: 'mapOverlay',
+  });
+  const rootMapOverlayGroup = await models.mapOverlayGroup.findRootMapOverlayGroup();
+  const permittedMapOverlayGroupIds = findPermittedMapOverlayGroupIds(
+    relationByChildId,
+    permittedMapOverlayGroupRelations,
+    rootMapOverlayGroup.id,
+  );
```

Comment:
I believe we don't need the `findPermittedMapOverlayGroupIds` method:

```suggestion
  const permittedMapOverlayGroupIds = getUniqueEntries(
    permittedMapOverlayGroupRelations.map(r => r.map_overlay_group_id),
  );
```

Reply:
Ah I had to do it this way because we can have multi map overlay groups, just getting the group ids from `permittedMapOverlayGroupRelations` is not enough (`permittedMapOverlayGroupRelations` will only contain the relations that have map overlay children). So I had to traverse up the list until it hits the root map overlay group

Reply:
You are right - I realised that later, after reading `findPermittedMapOverlayGroupIds`... Perhaps a one-line comment would help clarify